lib/resque/worker.rb

```diff
@@ -178,9 +178,15 @@ def prepare
     WILDCARDS = ['*', '?', '{', '}', '[', ']'].freeze
 
     def queues=(queues)
-      queues = queues.empty? ? (ENV["QUEUES"] || ENV['QUEUE']).to_s.split(',') : queues
-      @queues = queues.map { |queue| queue.to_s.strip }
-      @has_dynamic_queues = WILDCARDS.any? {|char| @queues.join.include?(char) }
+      queues = (ENV["QUEUES"] || ENV['QUEUE']).to_s.split(',') if queues.empty?
+      queues = queues.map { |queue| queue.to_s.strip }
+
+      @skip_queues, @queues = queues.partition { |queue| queue.start_with?('!') }
+      @skip_queues.map! { |queue| queue[1..-1] }
+
+      @has_dynamic_queues =
+        @skip_queues.any? || WILDCARDS.any? { |char| @queues.join.include?(char) }
```

Comment:
I'm not following the `@skip_queues.any?` condition here, could you explain? Why is this required?

Reply:
The behavior of `#queues` is dependent on the value of `@has_dynamic_queues`: if it's true, the method returns the result of filtering `@queues` with`#glob_match`; if it's false, the method returns `@queues` directly. Since `#glob_match` will cause skipped queues to be filtered out, we want to make sure it's called if we have  `@skip_queues.any?`.

https://github.com/resque/resque/blob/92a4564fac9af56ad0b1b2a37ad51dbea5c2bf36/lib/resque/worker.rb#L202-L215

It's like wildcards: just as we don't want to return `my-queue-*` from `#queues`, we don't want to return `!my-queue`, but rather the result of filtering `@queues` with the pattern.

Reply:
Got it, thanks for explaining!